code/modules/asset_cache/spritesheet/batched/universal_icon.dm

```diff
@@ -0,0 +1,374 @@
+
+/// Recursively converts all contained [/datum/universal_icon]s and their associated [/datum/icon_transformer]s into list form so the transforms can be JSON encoded.
+/datum/icon_transformer/proc/to_list()
+	RETURN_TYPE(/list)
+	var/list/transforms_out = list()
+	var/list/transforms_original = src.transforms.Copy()
+	for(var/list/transform as anything in transforms_original)
+		var/list/this_transform = transform.Copy() // copy it so we don't mutate the original
+		if(transform["type"] == RUSTG_ICONFORGE_BLEND_ICON)
+			var/datum/universal_icon/icon_object = this_transform["icon"]
+			if(!istype(icon_object))
+				stack_trace("Invalid icon found in icon transformer during to_list()! [icon_object]")
+				continue
+			// This mutates the inner transform list!!! Make sure it only runs on copies.
+			this_transform["icon"] = icon_object.to_list()
+		transforms_out += list(this_transform)
+	return transforms_out
+
+/// Reverse operation of /datum/icon_transformer/to_list()
+/proc/icon_transformer_from_list(list/input)
+	RETURN_TYPE(/datum/icon_transformer)
+	var/list/transforms = list()
+	for(var/list/transform as anything in input)
+		// don't mutate the input :(
+		var/this_transform = transform.Copy()
+		if(transform["type"] == RUSTG_ICONFORGE_BLEND_ICON)
+			this_transform["icon"] = universal_icon_from_list(transform["icon"])
+		transforms += list(this_transform)
+	var/datum/icon_transformer/transformer = new()
+	transformer.transforms = transforms
+	return transformer
+
+/// Constructs a transformer, with optional color multiply pre-added.
+/proc/color_transform(color=null)
+	RETURN_TYPE(/datum/icon_transformer)
+	var/datum/icon_transformer/transform = new()
+	if(color)
+		transform.blend_color(color, ICON_MULTIPLY)
+	return transform
+
+/// Converts a GAGS atom to a universal icon by generating blend operations.
+/proc/gags_to_universal_icon(atom/path)
+	RETURN_TYPE(/datum/universal_icon)
+	if(!ispath(path, /atom) || !initial(path.greyscale_config) || !initial(path.greyscale_colors))
+		CRASH("gags_to_universal_icon() received an invalid path of \"[path]\"!")
+	var/datum/greyscale_config/config = initial(path.greyscale_config)
+	var/colors = initial(path.greyscale_colors)
+	var/datum/universal_icon/entry = SSgreyscale.GetColoredIconByType_universal_icon(config, colors, initial(path.icon_state))
+	return entry
+
+/// Gets the relevant universal icon for an atom, when displayed in TGUI. (see: icon_state_preview)
+/// Supports GAGS items and colored items.
+/proc/get_display_icon_for(atom/A)
+	if (!ispath(A, /atom))
+		return FALSE
+	var/icon_file = initial(A.icon)
+	var/icon_state = initial(A.icon_state)
+	if(initial(A.greyscale_config) && initial(A.greyscale_colors))
+		return gags_to_universal_icon(A)
+	if(ispath(A, /obj))
+		var/obj/I = A
+		if(initial(I.icon_state_preview))
+			icon_state = initial(I.icon_state_preview)
+	return uni_icon(icon_file, icon_state, color=initial(A.color))
```

Comment:
```suggestion
/proc/get_display_icon_for(atom/atom_path)
	if (!ispath(atom_path, /atom))
		return FALSE
	var/icon_file = initial(A.icon)
	var/icon_state = initial(atom_path.icon_state)
	if(initial(atom_path.greyscale_config) && initial(atom_path.greyscale_colors))
		return gags_to_universal_icon(atom_path)
	if(ispath(atom_path, /obj))
		var/obj/obj_path = A
		if(initial(obj_path.icon_state_preview))
			icon_state = initial(obj_path.icon_state_preview)
	return uni_icon(icon_file, icon_state, color=initial(atom_path.color))
```
As long as they're sufficiently descriptive, you can also name them to something else tbh